### DISCUSSION TITLE
Getting calculated distance from OrderByDistance

### DISCUSSION THREAD

**Discussion by georgiosd** [ravendb/ravendb#15104] [Q&A] [answered]
Hey guys

I remembered there was a way to get the calculated distance when doing spatial queries and couldn't find it in the docs but I was able to find past conversations on this for 4.x.

https://groups.google.com/g/ravendb/c/FFREycqmORM/m/pHB93ozSAwAJ
https://issues.hibernatingrhinos.com/issue/RavenDB-13682
https://github.com/ravendb/ravendb/pull/9367/commits

Was this never ported to 5.x? I couldn't get it work:

![image](https://user-images.githubusercontent.com/1672054/195293058-10dc9a7f-2b48-4029-add4-758d7aeecff4.png)

I also can't find a reference in the code: https://github.com/ravendb/ravendb/search?q=%40distance&type=code

**Comment by ayende** [comment]
You need to use:

```
                    Assert.Equal(48.99, Math.Round((double)metadata.GetObject("@spatial")["Distance"], 2));
```

So your query should be like this, I think.

```
   Distance: getMetadata(o)["@spatial"]["Distance"]
```

**Reply by georgiosd** [reply]
Nada :-\

![image](https://user-images.githubusercontent.com/1672054/195294473-449f1f5e-1751-47eb-831a-397865940461.png)

**Reply by georgiosd** [reply]
`@spatial` doesn't show up in current code: https://github.com/ravendb/ravendb/search?q=%40spatial&type=code

**Reply by ayende** [reply]
Using the sample data, run this query:

```
from "Employees"  as o
where o.Address.Location  != null
order by spatial.distance(
    spatial.point(o.Address.Location.Latitude, o.Address.Location.Longitude),
    spatial.point(47.6, -122.1)
)
```

You'll then see:

![image](https://user-images.githubusercontent.com/116915/195296194-c81bef17-f00c-41b8-8505-fd79c9175112.png)

`@spatial` is defined here: 
https://github.com/ravendb/ravendb/blob/939bc815cc674d7c78ad6e03efc8cb7ee393dde3/src/Raven.Client/Constants.cs#L229

And used here:
https://github.com/ravendb/ravendb/blob/55c2c81bd07393ac819dd0103680005a0a8f8c89/src/Raven.Server/Documents/Document.cs#L86

There is an issue with not exposing the `@spatial` to JavaScript, I'm afraid, see: 
https://issues.hibernatingrhinos.com/issue/RavenDB-19453/spatial-metadata-value-isnt-exposed-to-javascript-projections

If you'll skip the projection, you'll get it.

**Reply by ayende** [reply]
https://github.com/ravendb/ravendb/pull/15107

**Reply by georgiosd** [reply]
Hm, strange that Github wouldnt find this string. I guess they're not indexing with Raven :)

Have you found it to work in LINQ projections? I had trouble getting it with `(double?)((IMetadataDictionary)RavenQuery.Metadata(result)["@spatial"])["Distance"]`.

Perhaps it was because it was an index query? All examples I've seen so far operate on an collection query.

**Reply by ayende** [reply]
There isn't a difference between them.
What is the actual problem that you are getting there?

**Reply by georgiosd** [reply]
No result (null). I'll prep a repro.

**Reply by ayende** [reply]
Checking if you were able to reproduce?

**Reply by georgiosd** [reply]
See this is why I shouldn't let 2 weeks pass before attempting a repro... I am pretty sure I was getting zero values instead of an exception (`Member named @metadata does not exist`).

Alas, this is the repro with the same structure as the actual app. Obviously there are more maps in the original thing, hence the multi-map index.

I get that multi-map indexes dont associate with just one document, which means there is no metadata either - but then how to get the distance?

```
async Task Main()
{
	var store = new DocumentStore
	{
		Urls = new[] { "http://localhost:8080" },
		Database = "test"
	};
	
	store.Initialize();
	
	store.ExecuteIndex(new RestoIndex());

	using (var session = store.OpenAsyncSession())
	{
		await session.StoreAsync(new Resto
		{
			Lat = 25.0676339,
			Lng = 55.141291
		});
		
		await session.SaveChangesAsync();
	}
	
	using(var session = store.OpenAsyncSession())
	{
		var query = session.Query<RestoIndex.Result, RestoIndex>()
			.Spatial(x => x.Location, c => c.WithinRadius(100000, 25, 55));
			
		(await query.ToArrayAsync()).Dump();
			
		var projection = from result in query
			select new
			{
				Distance = (double?)((IMetadataDictionary)RavenQuery.Metadata(result)["@spatial"])["Distance"]
			};
		
		(await projection.ToArrayAsync()).Dump();
	}
}

class Resto
{
	public string Id { get; set; }
	
	public double Lat { get; set; }
	
	public double Lng { get; set; }
}

public class RestoIndex : AbstractMultiMapIndexCreationTask<RestoIndex.Result>
{
	public class Result
	{
		public string Id { get; set; } = null!;

		public double LocationLatitude { get; set; }

		public double LocationLongitude { get; set; }

		public object? Location { get; set; }
	}

	public RestoIndex()
	{
		AddMap<Resto>(restaurants => from restaurant in restaurants
										  select new Result
										  {
										  		Id = restaurant.Id,
											  LocationLatitude = (double)restaurant.Lat,
											  LocationLongitude = (double)restaurant.Lng,
											  Location = null,
										  });

		Reduce = results => from result in results
							group result by result.Id
			into g
							let restaurant = g.FirstOrDefault()
							select new Result
							{
								Id = g.Key,
								LocationLatitude = restaurant.LocationLatitude,
								LocationLongitude = restaurant.LocationLongitude,
								Location = CreateSpatialField(restaurant.LocationLatitude, restaurant.LocationLongitude)
							};
							
		StoreAllFields(FieldStorage.Yes);
	}

}
```
It still fails on a single-map index but with no exception on the metadata, just null result:

```
async Task Main()
{
	var store = new DocumentStore
	{
		Urls = new[] { "http://localhost:8080" },
		Database = "test"
	};
	
	store.Initialize();
	
	//store.ExecuteIndex(new RestoIndex());
//
//	using (var session = store.OpenAsyncSession())
//	{
//		await session.StoreAsync(new Resto
//		{
//			Lat = 25.0676339,
//			Lng = 55.141291
//		});
//		
//		await session.SaveChangesAsync();
//	}
	
	using(var session = store.OpenAsyncSession())
	{
		var query = session.Query<RestoIndex.Result, RestoIndex>()
			.Spatial(x => x.Location, c => c.WithinRadius(100000, 25, 55));
			
		var projection = from result in query
			select new
			{
				Distance = (double?)((IMetadataDictionary)RavenQuery.Metadata(result)["@spatial"])["Distance"]
			};
		
		(await projection.ToArrayAsync()).Dump();
	}
}

public class Resto
{
	public string Id { get; set; }
	
	public double Lat { get; set; }
	
	public double Lng { get; set; }
}

public class RestoIndex : AbstractIndexCreationTask<Resto, RestoIndex.Result>
{
	public class Result
	{
		public string Id { get; set; } = null!;

		public object? Location { get; set; }
	}

	public RestoIndex()
	{
		Map = restaurants => from restaurant in restaurants
										  select new Result
										  {
										  		Id = restaurant.Id,
											  Location = CreateSpatialField(restaurant.Lat, restaurant.Lng)
										  };
	}
}
```

**Reply by garayx** [reply]
why would you expect to have `["@spatial"]["Distance"]` in metadata if you don't call `OrderByDistance()` or `spatial.distance()` ?
In any case I get `@metadata does not exist` I guess thats becasue query on index returns a document as result, so I converted your query to collection query and got the desired results:


```
from Restos as i
where spatial.within(spatial.point(i.Lat,i.Lng), spatial.circle($p0, $p1, $p2)) 
order by spatial.distance(
    spatial.point(i.Lat,i.Lng),
    spatial.point(25,55)
)
select { Distance: getMetadata(i)['@spatial'].Distance }
{"p0":100000.0,"p1":25.0,"p2":55.0}
```

**Reply by georgiosd** [reply]
You're right, my original code had `OrderByDistance` but I failed to add to the repro. However it still doesn't work an on index result, whether single or multi map.

Is it only supported on collection queries then?

Updated single-map index repro:

```
async Task Main()
{
	var store = new DocumentStore
	{
		Urls = new[] { "http://localhost:8080" },
		Database = "test"
	};
	
	store.Initialize();
	
	store.ExecuteIndex(new RestoIndex());

	using (var session = store.OpenAsyncSession())
	{
		await session.StoreAsync(new Resto
		{
			Lat = 25.0676339,
			Lng = 55.141291
		});
		
		await session.SaveChangesAsync();
	}
	
	using(var session = store.OpenAsyncSession())
	{
		var query = session.Query<RestoIndex.Result, RestoIndex>()
			.OrderByDistance(x => x.Location, 25, 55);
			
		var projection = from result in query
			select new
			{
				Distance = (double?)((IMetadataDictionary)RavenQuery.Metadata(result)["@spatial"])["Distance"]
			};
		
		(await projection.ToArrayAsync()).Dump();
	}
}

public class Resto
{
	public string Id { get; set; }
	
	public double Lat { get; set; }
	
	public double Lng { get; set; }
}

public class RestoIndex : AbstractIndexCreationTask<Resto, RestoIndex.Result>
{
	public class Result
	{
		public string Id { get; set; } = null!;

		public object? Location { get; set; }
	}

	public RestoIndex()
	{
		Map = restaurants => from restaurant in restaurants
										  select new Result
										  {
										  		Id = restaurant.Id,
											  Location = CreateSpatialField(restaurant.Lat, restaurant.Lng)
										  };
	}
}
```

**Reply by garayx** [reply]
yes, please see the query in the previous comment, it uses collection query and return the result as expected

**Comment by georgiosd** [comment]
I apologize for the delay, it's still on my TODO.

**Comment by georgiosd** [comment]
Ok, but what we have is an index query so a collection query won't do very
much good.

Do you plan to add support for this on indexes?

**Reply by ayende** [reply]
This *should* work

**Reply by garayx** [reply]
opened an issue to check it
https://issues.hibernatingrhinos.com/issue/RavenDB-19544/Check-why-spatial-query-with-distance-doesnt-work-on-index-query

**Comment by ayende** [comment]
Following up on that, the underlying issue isn't about auto indexes vs. static indexes.
The problem is that you are using a map/reduce index, and that one _doesn't_ have metadata associated with it.

**Reply by georgiosd** [reply]
I get that, and thank you for fixing it. However, in [my last example](https://github.com/ravendb/ravendb/discussions/15104#discussioncomment-4115652), it's a simple map index and I still get null. Is my code wrong?

**Reply by ayende** [reply]
Hm... you are correct, not sure about that. I added a test to verify that it works, but I didn't bother to figure out why it failed in that scenario separate from the map/reduce one.

**Comment by ayende** [comment, marked as answer]
This is fixed here: https://github.com/ravendb/ravendb/pull/15278

**Comment by georgiosd** [comment]
If it fixes both scenarios I'm happy :)
